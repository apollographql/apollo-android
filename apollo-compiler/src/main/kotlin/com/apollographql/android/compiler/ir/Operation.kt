package com.apollographql.android.compiler.ir

import com.apollographql.android.compiler.SchemaTypeSpecBuilder
import com.apollographql.apollo.api.Operation
import com.squareup.javapoet.TypeSpec
import javax.lang.model.element.Modifier

data class Operation(
    val operationName: String,
    val operationType: String,
    val variables: List<Variable>,
    val source: String,
    val fields: List<Field>,
    val filePath: String,
    val fragmentsReferenced: List<String>
) : CodeGenerator {
  override fun toTypeSpec(context: CodeGenerationContext): TypeSpec =
      SchemaTypeSpecBuilder(DATA_TYPE_NAME, fields, emptyList(), emptyList(), context)
          .build(Modifier.PUBLIC, Modifier.STATIC)
          .toBuilder()
          .addSuperinterface(Operation.Data::class.java)
          .build()

  companion object {
    val DATA_TYPE_NAME = "Data"
  }
}
