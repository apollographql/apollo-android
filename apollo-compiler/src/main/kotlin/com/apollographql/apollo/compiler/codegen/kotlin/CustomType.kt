package com.apollographql.apollo.compiler.codegen.kotlin

import com.apollographql.apollo.api.ScalarType
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy

fun KotlinCodeGen.customScalarTypeSpec(customTypes: Map<String, String>): TypeSpec {
  return TypeSpec.enumBuilder("CustomType")
      .addAnnotation(generatedByApolloAnnotation)
      .addSuperinterface(ScalarType::class.java)
      .apply {
        customTypes.map { (schemaType, customType) ->
          addEnumConstant(
              name = schemaType.normalizeGraphQLType().toUpperCase(),
              typeSpec = enumConstantTypeSpec(schemaType = schemaType, customType = customType)
          )
        }
      }
      .build()
}

private fun enumConstantTypeSpec(schemaType: String, customType: String): TypeSpec {
  return TypeSpec.anonymousClassBuilder()
      .addFunction(FunSpec.builder("typeName")
          .addModifiers(KModifier.OVERRIDE)
          .returns(String::class)
          .addStatement("return %S", schemaType)
          .build()
      )
      .addFunction(FunSpec.builder("javaType")
          .returns(Class::class.asClassName().parameterizedBy(STAR))
          .addModifiers(KModifier.OVERRIDE)
          .addStatement("return %L::class.java", customType)
          .build()
      )
      .build()
}