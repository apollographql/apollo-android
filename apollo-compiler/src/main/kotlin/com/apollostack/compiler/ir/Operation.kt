package com.apollostack.compiler.ir

class Operation(val operationName: String, val operationType: String, val variables: List<String>,
    val source: String, val fields: List<Field>) {

}