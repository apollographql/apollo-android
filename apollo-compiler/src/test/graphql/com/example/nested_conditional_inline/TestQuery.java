package com.example.nested_conditional_inline;

import com.apollographql.apollo.api.Field;
import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.api.Query;
import com.apollographql.apollo.api.ResponseFieldMapper;
import com.apollographql.apollo.api.ResponseReader;
import com.apollographql.apollo.api.internal.Optional;
import com.apollographql.apollo.api.internal.UnmodifiableMapBuilder;
import com.example.nested_conditional_inline.type.Episode;
import java.io.IOException;
import java.lang.Double;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import javax.annotation.Generated;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;

@Generated("Apollo GraphQL")
public final class TestQuery implements Query<TestQuery.Data, Optional<TestQuery.Data>, TestQuery.Variables> {
  public static final String OPERATION_DEFINITION = "query TestQuery($episode: Episode) {\n"
      + "  hero(episode: $episode) {\n"
      + "    __typename\n"
      + "    name\n"
      + "    ... on Human {\n"
      + "      __typename\n"
      + "      friends {\n"
      + "        __typename\n"
      + "        name\n"
      + "        ... on Human {\n"
      + "          __typename\n"
      + "          height(unit: FOOT)\n"
      + "        }\n"
      + "      }\n"
      + "    }\n"
      + "    ... on Droid {\n"
      + "      __typename\n"
      + "      friends {\n"
      + "        __typename\n"
      + "        name\n"
      + "        ... on Human {\n"
      + "          __typename\n"
      + "          height(unit: METER)\n"
      + "        }\n"
      + "      }\n"
      + "    }\n"
      + "  }\n"
      + "}";

  public static final String QUERY_DOCUMENT = OPERATION_DEFINITION;

  private final TestQuery.Variables variables;

  public TestQuery(@Nullable Episode episode) {
    variables = new TestQuery.Variables(episode);
  }

  @Override
  public String queryDocument() {
    return QUERY_DOCUMENT;
  }

  @Override
  public Optional<TestQuery.Data> wrapData(TestQuery.Data data) {
    return Optional.fromNullable(data);
  }

  @Override
  public TestQuery.Variables variables() {
    return variables;
  }

  @Override
  public ResponseFieldMapper<TestQuery.Data> responseFieldMapper() {
    return new Data.Mapper();
  }

  public static Builder builder() {
    return new Builder();
  }

  public static final class Variables extends Operation.Variables {
    private final @Nullable Episode episode;

    private final transient Map<String, Object> valueMap = new LinkedHashMap<>();

    Variables(@Nullable Episode episode) {
      this.episode = episode;
      this.valueMap.put("episode", episode);
    }

    public @Nullable Episode episode() {
      return episode;
    }

    @Override
    public Map<String, Object> valueMap() {
      return Collections.unmodifiableMap(valueMap);
    }
  }

  public static final class Builder {
    private @Nullable Episode episode;

    Builder() {
    }

    public Builder episode(@Nullable Episode episode) {
      this.episode = episode;
      return this;
    }

    public TestQuery build() {
      return new TestQuery(episode);
    }
  }

  public static class Data implements Operation.Data {
    private final Optional<Hero> hero;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public Data(@Nullable Hero hero) {
      this.hero = Optional.fromNullable(hero);
    }

    public Optional<Hero> hero() {
      return this.hero;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "Data{"
          + "hero=" + hero
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof Data) {
        Data that = (Data) o;
        return ((this.hero == null) ? (that.hero == null) : this.hero.equals(that.hero));
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= (hero == null) ? 0 : hero.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<Data> {
      final Hero.Mapper heroFieldMapper = new Hero.Mapper();

      final Field[] fields = {
        Field.forObject("hero", "hero", new UnmodifiableMapBuilder<String, Object>(1)
          .put("episode", new UnmodifiableMapBuilder<String, Object>(2)
            .put("kind", "Variable")
            .put("variableName", "episode")
          .build())
        .build(), true, new Field.ObjectReader<Hero>() {
          @Override public Hero read(final ResponseReader reader) throws IOException {
            return heroFieldMapper.map(reader);
          }
        })
      };

      @Override
      public Data map(ResponseReader reader) throws IOException {
        final Hero hero = reader.read(fields[0]);
        return new Data(hero);
      }
    }
  }

  public static class Hero {
    private final @Nonnull String __typename;

    private final @Nonnull String name;

    private final Optional<AsHuman> asHuman;

    private final Optional<AsDroid> asDroid;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public Hero(@Nonnull String __typename, @Nonnull String name, @Nullable AsHuman asHuman,
        @Nullable AsDroid asDroid) {
      this.__typename = __typename;
      this.name = name;
      this.asHuman = Optional.fromNullable(asHuman);
      this.asDroid = Optional.fromNullable(asDroid);
    }

    public @Nonnull String __typename() {
      return this.__typename;
    }

    public @Nonnull String name() {
      return this.name;
    }

    public Optional<AsHuman> asHuman() {
      return this.asHuman;
    }

    public Optional<AsDroid> asDroid() {
      return this.asDroid;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "Hero{"
          + "__typename=" + __typename + ", "
          + "name=" + name + ", "
          + "asHuman=" + asHuman + ", "
          + "asDroid=" + asDroid
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof Hero) {
        Hero that = (Hero) o;
        return ((this.__typename == null) ? (that.__typename == null) : this.__typename.equals(that.__typename))
         && ((this.name == null) ? (that.name == null) : this.name.equals(that.name))
         && ((this.asHuman == null) ? (that.asHuman == null) : this.asHuman.equals(that.asHuman))
         && ((this.asDroid == null) ? (that.asDroid == null) : this.asDroid.equals(that.asDroid));
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= (__typename == null) ? 0 : __typename.hashCode();
        h *= 1000003;
        h ^= (name == null) ? 0 : name.hashCode();
        h *= 1000003;
        h ^= (asHuman == null) ? 0 : asHuman.hashCode();
        h *= 1000003;
        h ^= (asDroid == null) ? 0 : asDroid.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<Hero> {
      final AsHuman.Mapper asHumanFieldMapper = new AsHuman.Mapper();

      final AsDroid.Mapper asDroidFieldMapper = new AsDroid.Mapper();

      final Field[] fields = {
        Field.forString("__typename", "__typename", null, false),
        Field.forString("name", "name", null, false),
        Field.forConditionalType("__typename", "__typename", new Field.ConditionalTypeReader<AsHuman>() {
          @Override
          public AsHuman read(String conditionalType, ResponseReader reader) throws IOException {
            if (conditionalType.equals("Human")) {
              return asHumanFieldMapper.map(reader);
            } else {
              return null;
            }
          }
        }),
        Field.forConditionalType("__typename", "__typename", new Field.ConditionalTypeReader<AsDroid>() {
          @Override
          public AsDroid read(String conditionalType, ResponseReader reader) throws IOException {
            if (conditionalType.equals("Droid")) {
              return asDroidFieldMapper.map(reader);
            } else {
              return null;
            }
          }
        })
      };

      @Override
      public Hero map(ResponseReader reader) throws IOException {
        final String __typename = reader.read(fields[0]);
        final String name = reader.read(fields[1]);
        final AsHuman asHuman = reader.read(fields[2]);
        final AsDroid asDroid = reader.read(fields[3]);
        return new Hero(__typename, name, asHuman, asDroid);
      }
    }
  }

  public static class AsHuman {
    private final @Nonnull String __typename;

    private final @Nonnull String name;

    private final Optional<List<Friend>> friends;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public AsHuman(@Nonnull String __typename, @Nonnull String name,
        @Nullable List<Friend> friends) {
      this.__typename = __typename;
      this.name = name;
      this.friends = Optional.fromNullable(friends);
    }

    public @Nonnull String __typename() {
      return this.__typename;
    }

    public @Nonnull String name() {
      return this.name;
    }

    public Optional<List<Friend>> friends() {
      return this.friends;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "AsHuman{"
          + "__typename=" + __typename + ", "
          + "name=" + name + ", "
          + "friends=" + friends
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof AsHuman) {
        AsHuman that = (AsHuman) o;
        return ((this.__typename == null) ? (that.__typename == null) : this.__typename.equals(that.__typename))
         && ((this.name == null) ? (that.name == null) : this.name.equals(that.name))
         && ((this.friends == null) ? (that.friends == null) : this.friends.equals(that.friends));
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= (__typename == null) ? 0 : __typename.hashCode();
        h *= 1000003;
        h ^= (name == null) ? 0 : name.hashCode();
        h *= 1000003;
        h ^= (friends == null) ? 0 : friends.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<AsHuman> {
      final Friend.Mapper friendFieldMapper = new Friend.Mapper();

      final Field[] fields = {
        Field.forString("__typename", "__typename", null, false),
        Field.forString("name", "name", null, false),
        Field.forList("friends", "friends", null, true, new Field.ObjectReader<Friend>() {
          @Override public Friend read(final ResponseReader reader) throws IOException {
            return friendFieldMapper.map(reader);
          }
        })
      };

      @Override
      public AsHuman map(ResponseReader reader) throws IOException {
        final String __typename = reader.read(fields[0]);
        final String name = reader.read(fields[1]);
        final List<Friend> friends = reader.read(fields[2]);
        return new AsHuman(__typename, name, friends);
      }
    }
  }

  public static class Friend {
    private final @Nonnull String __typename;

    private final @Nonnull String name;

    private final Optional<AsHuman1> asHuman;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public Friend(@Nonnull String __typename, @Nonnull String name, @Nullable AsHuman1 asHuman) {
      this.__typename = __typename;
      this.name = name;
      this.asHuman = Optional.fromNullable(asHuman);
    }

    public @Nonnull String __typename() {
      return this.__typename;
    }

    public @Nonnull String name() {
      return this.name;
    }

    public Optional<AsHuman1> asHuman() {
      return this.asHuman;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "Friend{"
          + "__typename=" + __typename + ", "
          + "name=" + name + ", "
          + "asHuman=" + asHuman
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof Friend) {
        Friend that = (Friend) o;
        return ((this.__typename == null) ? (that.__typename == null) : this.__typename.equals(that.__typename))
         && ((this.name == null) ? (that.name == null) : this.name.equals(that.name))
         && ((this.asHuman == null) ? (that.asHuman == null) : this.asHuman.equals(that.asHuman));
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= (__typename == null) ? 0 : __typename.hashCode();
        h *= 1000003;
        h ^= (name == null) ? 0 : name.hashCode();
        h *= 1000003;
        h ^= (asHuman == null) ? 0 : asHuman.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<Friend> {
      final AsHuman1.Mapper asHuman1FieldMapper = new AsHuman1.Mapper();

      final Field[] fields = {
        Field.forString("__typename", "__typename", null, false),
        Field.forString("name", "name", null, false),
        Field.forConditionalType("__typename", "__typename", new Field.ConditionalTypeReader<AsHuman1>() {
          @Override
          public AsHuman1 read(String conditionalType, ResponseReader reader) throws IOException {
            if (conditionalType.equals("Human")) {
              return asHuman1FieldMapper.map(reader);
            } else {
              return null;
            }
          }
        })
      };

      @Override
      public Friend map(ResponseReader reader) throws IOException {
        final String __typename = reader.read(fields[0]);
        final String name = reader.read(fields[1]);
        final AsHuman1 asHuman = reader.read(fields[2]);
        return new Friend(__typename, name, asHuman);
      }
    }
  }

  public static class AsHuman1 {
    private final @Nonnull String __typename;

    private final @Nonnull String name;

    private final Optional<Double> height;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public AsHuman1(@Nonnull String __typename, @Nonnull String name, @Nullable Double height) {
      this.__typename = __typename;
      this.name = name;
      this.height = Optional.fromNullable(height);
    }

    public @Nonnull String __typename() {
      return this.__typename;
    }

    public @Nonnull String name() {
      return this.name;
    }

    public Optional<Double> height() {
      return this.height;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "AsHuman1{"
          + "__typename=" + __typename + ", "
          + "name=" + name + ", "
          + "height=" + height
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof AsHuman1) {
        AsHuman1 that = (AsHuman1) o;
        return ((this.__typename == null) ? (that.__typename == null) : this.__typename.equals(that.__typename))
         && ((this.name == null) ? (that.name == null) : this.name.equals(that.name))
         && ((this.height == null) ? (that.height == null) : this.height.equals(that.height));
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= (__typename == null) ? 0 : __typename.hashCode();
        h *= 1000003;
        h ^= (name == null) ? 0 : name.hashCode();
        h *= 1000003;
        h ^= (height == null) ? 0 : height.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<AsHuman1> {
      final Field[] fields = {
        Field.forString("__typename", "__typename", null, false),
        Field.forString("name", "name", null, false),
        Field.forDouble("height", "height", new UnmodifiableMapBuilder<String, Object>(1)
          .put("unit", "FOOT")
        .build(), true)
      };

      @Override
      public AsHuman1 map(ResponseReader reader) throws IOException {
        final String __typename = reader.read(fields[0]);
        final String name = reader.read(fields[1]);
        final Double height = reader.read(fields[2]);
        return new AsHuman1(__typename, name, height);
      }
    }
  }

  public static class AsDroid {
    private final @Nonnull String __typename;

    private final @Nonnull String name;

    private final Optional<List<Friend1>> friends;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public AsDroid(@Nonnull String __typename, @Nonnull String name,
        @Nullable List<Friend1> friends) {
      this.__typename = __typename;
      this.name = name;
      this.friends = Optional.fromNullable(friends);
    }

    public @Nonnull String __typename() {
      return this.__typename;
    }

    public @Nonnull String name() {
      return this.name;
    }

    public Optional<List<Friend1>> friends() {
      return this.friends;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "AsDroid{"
          + "__typename=" + __typename + ", "
          + "name=" + name + ", "
          + "friends=" + friends
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof AsDroid) {
        AsDroid that = (AsDroid) o;
        return ((this.__typename == null) ? (that.__typename == null) : this.__typename.equals(that.__typename))
         && ((this.name == null) ? (that.name == null) : this.name.equals(that.name))
         && ((this.friends == null) ? (that.friends == null) : this.friends.equals(that.friends));
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= (__typename == null) ? 0 : __typename.hashCode();
        h *= 1000003;
        h ^= (name == null) ? 0 : name.hashCode();
        h *= 1000003;
        h ^= (friends == null) ? 0 : friends.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<AsDroid> {
      final Friend1.Mapper friend1FieldMapper = new Friend1.Mapper();

      final Field[] fields = {
        Field.forString("__typename", "__typename", null, false),
        Field.forString("name", "name", null, false),
        Field.forList("friends", "friends", null, true, new Field.ObjectReader<Friend1>() {
          @Override public Friend1 read(final ResponseReader reader) throws IOException {
            return friend1FieldMapper.map(reader);
          }
        })
      };

      @Override
      public AsDroid map(ResponseReader reader) throws IOException {
        final String __typename = reader.read(fields[0]);
        final String name = reader.read(fields[1]);
        final List<Friend1> friends = reader.read(fields[2]);
        return new AsDroid(__typename, name, friends);
      }
    }
  }

  public static class Friend1 {
    private final @Nonnull String __typename;

    private final @Nonnull String name;

    private final Optional<AsHuman2> asHuman;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public Friend1(@Nonnull String __typename, @Nonnull String name, @Nullable AsHuman2 asHuman) {
      this.__typename = __typename;
      this.name = name;
      this.asHuman = Optional.fromNullable(asHuman);
    }

    public @Nonnull String __typename() {
      return this.__typename;
    }

    public @Nonnull String name() {
      return this.name;
    }

    public Optional<AsHuman2> asHuman() {
      return this.asHuman;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "Friend1{"
          + "__typename=" + __typename + ", "
          + "name=" + name + ", "
          + "asHuman=" + asHuman
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof Friend1) {
        Friend1 that = (Friend1) o;
        return ((this.__typename == null) ? (that.__typename == null) : this.__typename.equals(that.__typename))
         && ((this.name == null) ? (that.name == null) : this.name.equals(that.name))
         && ((this.asHuman == null) ? (that.asHuman == null) : this.asHuman.equals(that.asHuman));
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= (__typename == null) ? 0 : __typename.hashCode();
        h *= 1000003;
        h ^= (name == null) ? 0 : name.hashCode();
        h *= 1000003;
        h ^= (asHuman == null) ? 0 : asHuman.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<Friend1> {
      final AsHuman2.Mapper asHuman2FieldMapper = new AsHuman2.Mapper();

      final Field[] fields = {
        Field.forString("__typename", "__typename", null, false),
        Field.forString("name", "name", null, false),
        Field.forConditionalType("__typename", "__typename", new Field.ConditionalTypeReader<AsHuman2>() {
          @Override
          public AsHuman2 read(String conditionalType, ResponseReader reader) throws IOException {
            if (conditionalType.equals("Human")) {
              return asHuman2FieldMapper.map(reader);
            } else {
              return null;
            }
          }
        })
      };

      @Override
      public Friend1 map(ResponseReader reader) throws IOException {
        final String __typename = reader.read(fields[0]);
        final String name = reader.read(fields[1]);
        final AsHuman2 asHuman = reader.read(fields[2]);
        return new Friend1(__typename, name, asHuman);
      }
    }
  }

  public static class AsHuman2 {
    private final @Nonnull String __typename;

    private final @Nonnull String name;

    private final Optional<Double> height;

    private volatile String $toString;

    private volatile int $hashCode;

    private volatile boolean $hashCodeMemoized;

    public AsHuman2(@Nonnull String __typename, @Nonnull String name, @Nullable Double height) {
      this.__typename = __typename;
      this.name = name;
      this.height = Optional.fromNullable(height);
    }

    public @Nonnull String __typename() {
      return this.__typename;
    }

    public @Nonnull String name() {
      return this.name;
    }

    public Optional<Double> height() {
      return this.height;
    }

    @Override
    public String toString() {
      if ($toString == null) {
        $toString = "AsHuman2{"
          + "__typename=" + __typename + ", "
          + "name=" + name + ", "
          + "height=" + height
          + "}";
      }
      return $toString;
    }

    @Override
    public boolean equals(Object o) {
      if (o == this) {
        return true;
      }
      if (o instanceof AsHuman2) {
        AsHuman2 that = (AsHuman2) o;
        return ((this.__typename == null) ? (that.__typename == null) : this.__typename.equals(that.__typename))
         && ((this.name == null) ? (that.name == null) : this.name.equals(that.name))
         && ((this.height == null) ? (that.height == null) : this.height.equals(that.height));
      }
      return false;
    }

    @Override
    public int hashCode() {
      if (!$hashCodeMemoized) {
        int h = 1;
        h *= 1000003;
        h ^= (__typename == null) ? 0 : __typename.hashCode();
        h *= 1000003;
        h ^= (name == null) ? 0 : name.hashCode();
        h *= 1000003;
        h ^= (height == null) ? 0 : height.hashCode();
        $hashCode = h;
        $hashCodeMemoized = true;
      }
      return $hashCode;
    }

    public static final class Mapper implements ResponseFieldMapper<AsHuman2> {
      final Field[] fields = {
        Field.forString("__typename", "__typename", null, false),
        Field.forString("name", "name", null, false),
        Field.forDouble("height", "height", new UnmodifiableMapBuilder<String, Object>(1)
          .put("unit", "METER")
        .build(), true)
      };

      @Override
      public AsHuman2 map(ResponseReader reader) throws IOException {
        final String __typename = reader.read(fields[0]);
        final String name = reader.read(fields[1]);
        final Double height = reader.read(fields[2]);
        return new AsHuman2(__typename, name, height);
      }
    }
  }
}
