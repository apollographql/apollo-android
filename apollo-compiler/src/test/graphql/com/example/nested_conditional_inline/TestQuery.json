{
	"operations": [
		{
			"filePath": "src/test/graphql/com/example/nested_conditional_inline/TestQuery.graphql",
			"operationName": "TestQuery",
			"operationType": "query",
			"variables": [
				{
					"name": "episode",
					"type": "Episode"
				}
			],
			"source": "query TestQuery($episode: Episode) {\n  hero(episode: $episode) {\n    __typename\n    name\n    ... on Human {\n      __typename\n      friends {\n        __typename\n        name\n        ... on Human {\n          __typename\n          height(unit: FOOT)\n        }\n      }\n    }\n    ... on Droid {\n      __typename\n      friends {\n        __typename\n        name\n        ... on Human {\n          __typename\n          height(unit: METER)\n        }\n      }\n    }\n  }\n}",
			"fields": [
				{
					"responseName": "hero",
					"fieldName": "hero",
					"type": "Character",
					"args": [
						{
							"name": "episode",
							"value": {
								"kind": "Variable",
								"variableName": "episode"
							}
						}
					],
					"fields": [
						{
							"responseName": "name",
							"fieldName": "name",
							"type": "String!"
						}
					],
					"fragmentSpreads": [],
					"inlineFragments": [
						{
							"typeCondition": "Human",
							"fields": [
								{
									"responseName": "name",
									"fieldName": "name",
									"type": "String!"
								},
								{
									"responseName": "friends",
									"fieldName": "friends",
									"type": "[Character]",
									"fields": [
										{
											"responseName": "name",
											"fieldName": "name",
											"type": "String!"
										}
									],
									"fragmentSpreads": [],
									"inlineFragments": [
										{
											"typeCondition": "Human",
											"fields": [
												{
													"responseName": "name",
													"fieldName": "name",
													"type": "String!"
												},
												{
													"responseName": "height",
													"fieldName": "height",
													"type": "Float",
													"args": [
														{
															"name": "unit",
															"value": "FOOT"
														}
													]
												}
											],
											"fragmentSpreads": []
										}
									]
								}
							],
							"fragmentSpreads": []
						},
						{
							"typeCondition": "Droid",
							"fields": [
								{
									"responseName": "name",
									"fieldName": "name",
									"type": "String!"
								},
								{
									"responseName": "friends",
									"fieldName": "friends",
									"type": "[Character]",
									"fields": [
										{
											"responseName": "name",
											"fieldName": "name",
											"type": "String!"
										}
									],
									"fragmentSpreads": [],
									"inlineFragments": [
										{
											"typeCondition": "Human",
											"fields": [
												{
													"responseName": "name",
													"fieldName": "name",
													"type": "String!"
												},
												{
													"responseName": "height",
													"fieldName": "height",
													"type": "Float",
													"args": [
														{
															"name": "unit",
															"value": "METER"
														}
													]
												}
											],
											"fragmentSpreads": []
										}
									]
								}
							],
							"fragmentSpreads": []
						}
					]
				}
			],
			"fragmentsReferenced": []
		}
	],
	"fragments": [],
	"typesUsed": [
		{
			"kind": "EnumType",
			"name": "Episode",
			"description": "The episodes in the Star Wars trilogy",
			"values": [
				{
					"name": "NEWHOPE",
					"description": "Star Wars Episode IV: A New Hope, released in 1977."
				},
				{
					"name": "EMPIRE",
					"description": "Star Wars Episode V: The Empire Strikes Back, released in 1980."
				},
				{
					"name": "JEDI",
					"description": "Star Wars Episode VI: Return of the Jedi, released in 1983."
				}
			]
		}
	]
}