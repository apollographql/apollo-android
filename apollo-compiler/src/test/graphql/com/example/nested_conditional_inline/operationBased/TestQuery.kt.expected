//
// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL version '$VERSION'.
//
package com.example.nested_conditional_inline

import com.apollographql.apollo3.api.Adapter
import com.apollographql.apollo3.api.CustomScalarAdapters
import com.apollographql.apollo3.api.FieldSet
import com.apollographql.apollo3.api.Query
import com.apollographql.apollo3.api.json.JsonWriter
import com.apollographql.apollo3.api.obj
import com.example.nested_conditional_inline.adapter.TestQuery_ResponseAdapter
import com.example.nested_conditional_inline.adapter.TestQuery_VariablesAdapter
import com.example.nested_conditional_inline.responsefields.TestQuery_ResponseFields
import com.example.nested_conditional_inline.type.Episode
import kotlin.Double
import kotlin.String
import kotlin.collections.List

data class TestQuery(
  val episode: Episode?
) : Query<TestQuery.Data> {
  override fun id(): String = OPERATION_ID

  override fun document(): String = OPERATION_DOCUMENT

  override fun name(): String = OPERATION_NAME

  override fun serializeVariables(writer: JsonWriter, customScalarAdapters: CustomScalarAdapters) {
    TestQuery_VariablesAdapter.toJson(writer, customScalarAdapters, this)
  }

  override fun adapter(): Adapter<Data> {
    return TestQuery_ResponseAdapter.Data.obj()
  }

  override fun fieldSets(): List<FieldSet> = TestQuery_ResponseFields.fields.first().fieldSets

  data class Data(
    val hero: Hero?
  ) : Query.Data {
    data class Hero(
      val __typename: String,
      /**
       * The name of the character
       */
      val name: String,
      /**
       * Synthetic field for inline fragment on Human
       */
      val onHuman: OnHuman?,
      /**
       * Synthetic field for inline fragment on Droid
       */
      val onDroid: OnDroid?
    ) {
      data class OnHuman(
        /**
         * This human's friends, or an empty list if they have none
         */
        val friends: List<Friend?>?
      ) {
        fun friendsFilterNotNull(): List<Friend>? = friends?.filterNotNull()

        data class Friend(
          val __typename: String,
          /**
           * The name of the character
           */
          val name: String,
          /**
           * Synthetic field for inline fragment on Human
           */
          val onHuman: OnHuman?
        ) {
          data class OnHuman(
            /**
             * Height in the preferred unit, default is meters
             */
            val height: Double?
          )
        }
      }

      data class OnDroid(
        /**
         * This droid's friends, or an empty list if they have none
         */
        val friends: List<Friend?>?
      ) {
        fun friendsFilterNotNull(): List<Friend>? = friends?.filterNotNull()

        data class Friend(
          val __typename: String,
          /**
           * The name of the character
           */
          val name: String,
          /**
           * Synthetic field for inline fragment on Human
           */
          val onHuman: OnHuman?
        ) {
          data class OnHuman(
            /**
             * Height in the preferred unit, default is meters
             */
            val height: Double?
          )
        }
      }
    }
  }

  companion object {
    const val OPERATION_ID: String =
        "a9f066a7d1092096ab154f16f32114a4bd71e959b789f37879249cdf6309ea86"

    /**
     * The minimized GraphQL document being sent to the server to save a few bytes.
     * The un-minimized version is:
     *
     * query TestQuery($episode: Episode) {
     *   hero(episode: $episode) {
     *     __typename
     *     name
     *     ... on Human {
     *       friends {
     *         __typename
     *         name
     *         ... on Human {
     *           height(unit: FOOT)
     *         }
     *       }
     *     }
     *     ... on Droid {
     *       friends {
     *         __typename
     *         name
     *         ... on Human {
     *           height(unit: METER)
     *         }
     *       }
     *     }
     *   }
     * }
     */
    const val OPERATION_DOCUMENT: String =
        "query TestQuery(${'$'}episode: Episode) { hero(episode: ${'$'}episode) { __typename name ... on Human { friends { __typename name ... on Human { height(unit: FOOT) } } } ... on Droid { friends { __typename name ... on Human { height(unit: METER) } } } } }"

    const val OPERATION_NAME: String = "TestQuery"
  }
}
